lib/ims/lti/models/messages/message.rb

```diff
@@ -73,20 +85,8 @@ def supers_params(instance_variable)
     add_required_params :lti_message_type, :lti_version
 
     def self.generate(params)
-      case params['lti_message_type']
-        when BasicLTILaunchRequest::MESSAGE_TYPE
-          BasicLTILaunchRequest.new(params)
-        when RegistrationRequest::MESSAGE_TYPE
-          RegistrationRequest.new(params)
-        when ContentItemSelectionRequest::MESSAGE_TYPE
-          ContentItemSelectionRequest.new(params)
-        when ContentItemSelection::MESSAGE_TYPE
-          ContentItemSelection.new(params)
-        when ToolProxyReregistrationRequest::MESSAGE_TYPE
-          ToolProxyReregistrationRequest.new(params)
-        else
-          self.new(params)
-      end
+      klass = self.descendants.select{|d| d::MESSAGE_TYPE == params['lti_message_type']}.first
+      klass ? klass.new(params) : self.new(params)
```

Comment:
I'm wondering if I should change the self.new on line 89 to Message.new. The side effect of the way it is now is that the default type created is whatever class you call it from, so if you call it from RequestMessage it will only return one of it's subtypes or itself. thoughts?